src/test/fuzz/versionbits.cpp

```diff
@@ -0,0 +1,343 @@
+    }
+
+    bool Condition(const CBlockIndex* pindex) const { return Condition(pindex->nVersion); }
+};
+
+/** Track blocks mined for test */
+class Blocks
+{
+private:
+    std::vector<CBlockIndex*> blocks;
+    const uint32_t m_start_time;
+    const int32_t m_signal;
+    const int32_t m_no_signal;
+
+public:
+    Blocks(uint32_t start_time, int32_t signal, int32_t no_signal) : m_start_time{start_time}, m_signal{signal}, m_no_signal{no_signal} { }
+
+    ~Blocks()
+    {
+        for (auto& v : blocks) {
+            delete v;
+        }
+        blocks.clear();
+    }
+
+    size_t size() const { return blocks.size(); }
+
+    CBlockIndex* tip() const
+    {
+        return blocks.empty() ? nullptr : blocks.back();
+    }
+
+    CBlockIndex* mine_block(bool signal)
+    {
+        CBlockHeader header;
+        header.nVersion = signal ? m_signal : m_no_signal;
+        header.nTime = m_start_time + blocks.size() * 600;
+        header.nBits = 0x1d00ffff;
+
+        CBlockIndex* current_block = new CBlockIndex{header};
+        current_block->pprev = tip();
+        current_block->nHeight = blocks.size();
+        current_block->BuildSkip();
+
+        blocks.push_back(current_block);
+
+        return current_block;
+    }
+};
+
+void initialize()
+{
+    return;
+}
+} // namespace
+
+FUZZ_TARGET_INIT(versionbits, initialize)
+{
+    FuzzedDataProvider fuzzed_data_provider(buffer.data(), buffer.size());
+
+    const int period = 32;
+    const int threshold = 29;
+
+    assert(0 < threshold && threshold <= period - 2);  // must be able to not signal!
+
+    const size_t max_periods = 16;
+    const size_t n_blocks = period * max_periods;
+
+    // pick the timestamp to switch based on a block
+    // note states will change *after* these blocks because mediantime lags
+    int start_block = fuzzed_data_provider.ConsumeIntegralInRange<int>(0, period * (max_periods-3));
+    int end_block = fuzzed_data_provider.ConsumeIntegralInRange<int>(start_block, period * (max_periods-3));
+
+    // between genesis and 2100-01-01
+    const int64_t block_start_time = fuzzed_data_provider.ConsumeIntegralInRange<uint32_t>(1231006505, 4102444800);
+
+    // too many blocks at 10min each might cause uint32_t time to overflow if
+    // block_start_time is at the end of the range above
+    assert(n_blocks < 320000);
+
+    // what values for version will we use to signal / not signal?
+    int32_t ver_signal = fuzzed_data_provider.ConsumeIntegral<int32_t>();
+    int32_t ver_nosignal = fuzzed_data_provider.ConsumeIntegral<int32_t>();
+
+    // select deployment parameters: bit, start time, timeout
+    int bit = fuzzed_data_provider.ConsumeIntegralInRange<int>(0, VERSIONBITS_NUM_BITS-1);
+
+    bool always_active_test = false;
+    bool never_active_test = false;
+    int64_t start_time;
+    int64_t timeout;
+    if (fuzzed_data_provider.ConsumeBool()) {
+        start_time = block_start_time + start_block*600;
```

Comment:
would it be possible to use params.consensus.nPowTargetSpacing instead of the hardcoded value?